tmp: start doc (#32)
DIRACGrid/diracx: docs/README.md

## Run a full diracx demo

This will allow you to run a demo setup.

The code changes will be reflected in the demo.

Requirement: docker, internet

```bash

> # Clone the diracx repository

Reviewer: Clone your fork ?

Reviewer 2: What if I just want to try the demo? (and not forking the whole project)

> 
> ```bash
> # Clone the diracx repository
> git clone git@github.com:<yourfork>/diracx.git
> 
> # Clone the diracx-chart repository
> git clone git@github.com:<yourfork>/diracx-charts.git
> 
> # Run the
> diracx-charts/run_demo.sh diracx/
> ```

Reviewer 3: ```suggestion

0. fork, in GitHub, the https://github.com/DIRACGrid/diracx/ and https://github.com/DIRACGrid/diracx-charts/ repositories.

```bash
git clone git@github.com:<yourfork>/diracx.git
cd diracx
git remote add upstream git@github.com:DIRACGrid/diracx.git
git fetch upstream

# Clone the diracx-chart repository
git clone git@github.com:<yourfork>/diracx-charts.git
cd diracx-charts
git remote add upstream git@github.com:DIRACGrid/diracx-charts.git
git fetch upstream

# Run the
diracx-charts/run_demo.sh diracx/
```
```

Author: I don't think this is necessary, it makes the instructions longer for no added value really. Plus, if they want to run a demo, they don't need to fork it.

> ```

Reviewer 4: Make the "login" process more explicit, by specifying that the email / password should be added in the identity provider page and not in the Swagger interface.

## Create the dev environment

This will help you setup a dev environment to run the unit tests

Requirements: conda, git

```bash

> # Clone the diracx repository
> git clone git@github.com:DIRACGrid/diracx.git

Reviewer 3: ```suggestion
```

cd diracx

# Create the mamba environment
mamba env create --file environment.yml
conda activate diracx-dev

# Make an editable installation of diracx

pip install -e .

# Install the patched DIRAC version
pip install git+https://github.com/chaen/DIRAC.git@chris-hack-a-ton

# Enable pre-commit
mamba install pre-commit
pre-commit install
```

### Run the test

Run the unit tests

```bash
# In the `diracx` folder

> pytest

Reviewer 3: ```suggestion
`pytest` for finding and running all tests. `pytest --regenerate-client tests/db` for running a subset of them. 
```

> mypy .
> pre-commit run --all-files

Reviewer 3: ```suggestion
`mypy .`
`pre-commit run --all-files`
```

Author: why ? They are in a code block

```

### Run a local instance of diracx

This only runs the diracx server, not any dependency like external IdP.

```bash
# In the `diracx` folder

> ./run_local.sh

Reviewer 5: In my case I should correct following in the script: 
tmp_dir="/Users/ruslan/BelleII/DIRACX/dev/tmp" #explicitly name tmp dir
export DIRACX_CONFIG_BACKEND_URL="git+file://${tmp_dir}/cs_store/initialRepo" #add this env var

### Databases

The `job_db` object is an instance of the given db (e.g. `JobDB`).
A single transaction is used for the duration of the request and it is automatically committed when the function returns.
If an exception is raised the transaction is rolled back.
Connections are pooled between requests.

### User information

The `user_info` object contains validation information about the current user, extracted from the authorization header of the request.
See the `UserInfo` class for the available properties.

## Adding a router

To add a router there are two steps:

1. Create a module in `diracx.routers` for the given service.
2. Add an entry to the `diracx.services` entrypoint.

We'll now make a `/parking/` router which contains information store in the `DummyDB`.

### Creating the router submodule

TODO: This isn't yet documented here however see `diracx.routers.well_known` for an example.

### Defining the new router in the `diracx.services` entrypoint

Modify the package's `setup.cfg`

```ini
[options.entry_points]
diracx.services =
	parking = diracx.routers.parking:router
```

> This will prefix the routes with `/parking/` and mark them with the `"parking"` tag in the OpenAPI spec.

Author: Add a line that it requires to reinstall diracx with pip to update this list

2. Add an entry to the `diracx.services` entrypoint.

We'll now make a `/parking/` router which contains information store in the `DummyDB`.

### Creating the router submodule

TODO: This isn't yet documented here however see `diracx.routers.well_known` for an example.

### Defining the new router in the `diracx.services` entrypoint

Modify the package's `setup.cfg`

```ini
[options.entry_points]
diracx.services =
	parking = diracx.routers.parking:router
```

This will prefix the routes with `/parking/` and mark them with the `"parking"` tag in the OpenAPI spec.

> [!WARNING]
> Any modification in the `setup.cfg` requires to re-install install `diracx`, even if it is a developer installation (`pip install -e`)

## Conventions

### Coding style

Most style are enforced with the pre-commit hooks which automatically reformat the code and sort the imports.
Code should follow PEP-8, particularly for [naming conventions](https://peps.python.org/pep-0008/#prescriptive-naming-conventions) as these aren't modified by the pre-commit hooks.

#### Import style

In addition to the automatic sorting

```python
from datetime import datetime, timedelta, timezone

import pytest
```

> 
> ### Databases
> 
> All SQL database interaction should use SQLAlchemy core and not the ORM to avoid performance issues.

Reviewer 6: This is already mentioned higher up

```suggestion
```
